src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -30,6 +34,71 @@ std::string toString(const Values & value)
     return fmt::format("({})", fmt::join(value, ", "));
 }
 
+bool isSafePrimaryDataKeyType(const IDataType & data_type)
+{
+    auto type_id = data_type.getTypeId();
+    switch (type_id)
+    {
+        case TypeIndex::Float32:
+        case TypeIndex::Float64:
+        case TypeIndex::Nullable:
+        case TypeIndex::Object:
+            return false;
+        case TypeIndex::Array:
+        {
+            const auto & data_type_array = static_cast<const DataTypeArray &>(data_type);
+            return isSafePrimaryDataKeyType(*data_type_array.getNestedType());
+        }
+        case TypeIndex::Tuple:
+        {
+            const auto & data_type_tuple = static_cast<const DataTypeTuple &>(data_type);
+            const auto & data_type_tuple_elements = data_type_tuple.getElements();
+            for (const auto & data_type_tuple_element : data_type_tuple_elements)
+                if (!isSafePrimaryDataKeyType(*data_type_tuple_element))
+                    return false;
+
+            return true;
+        }
+        case TypeIndex::LowCardinality:
+        {
+            const auto & data_type_low_cardinality = static_cast<const DataTypeLowCardinality &>(data_type);
+            return isSafePrimaryDataKeyType(*data_type_low_cardinality.getDictionaryType());
+        }
+        case TypeIndex::Map:
+        {
+            const auto & data_type_map = static_cast<const DataTypeMap &>(data_type);
+            return isSafePrimaryDataKeyType(*data_type_map.getKeyType()) && isSafePrimaryDataKeyType(*data_type_map.getValueType());
+        }
+        case TypeIndex::Variant:
+        {
+            const auto & data_type_variant = static_cast<const DataTypeVariant &>(data_type);
+            const auto & data_type_variant_elements = data_type_variant.getVariants();
+            for (const auto & data_type_variant_element : data_type_variant_elements)
+                if (!isSafePrimaryDataKeyType(*data_type_variant_element))
+                    return false;
+
+            return false;
+        }
+        default:
+        {
+            break;
+        }
+    }
+
+    return true;
+}
+
+bool isSafePrimaryKey(const KeyDescription & primary_key)
```

Comment:
maybe clarify in a comment, for what it should be safe

Reply:
Updated.